module/zfs/abd.c

```diff
@@ -0,0 +1,1113 @@
+ */
+unsigned int
+abd_scatter_bio_map_off(struct bio *bio, abd_t *abd, unsigned int bio_size,
+    size_t off)
+{
+	int i;
+	size_t len;
+	struct abd_miter aiter;
+
+	ABD_CHECK(abd);
+	ASSERT_ABD_SCATTER(abd);
+	ASSERT(bio_size <= abd->abd_size - off);
+
+	abd_miter_init(&aiter, abd, ABD_MITER_R);
+	abd_miter_advance(&aiter, off);
+
+	for (i = 0; i < bio->bi_max_vecs; i++) {
+		if (bio_size <= 0)
+			break;
+
+		len = MIN(bio_size, aiter.length);
+		ASSERT(len > 0);
+
+		if (bio_add_page(bio, sg_page(aiter.sg), len,
+		    aiter.offset) != len)
+			break;
+
+		bio_size -= len;
+		abd_miter_advance(&aiter, len);
+	}
+	return (bio_size);
+}
+
+/*
+ * bio_nr_pages for ABD.
+ * @off is the offset in @abd
+ */
+unsigned long
+abd_bio_nr_pages_off(abd_t *abd, unsigned int bio_size, size_t off)
+{
+	unsigned long pos = 0;
+	ABD_CHECK(abd);
+
+	if (ABD_IS_LINEAR(abd))
+		pos = (unsigned long)abd->abd_buf + off;
+	else
+		pos = abd->abd_offset + off;
+	return ((pos + bio_size + PAGE_SIZE-1)>>PAGE_SHIFT)-(pos>>PAGE_SHIFT);
+}
+#endif	/* _KERNEL */
+
+static inline abd_t *
+abd_alloc_struct(int nr_pages)
+{
+	abd_t *abd;
+	size_t asize = sizeof (abd_t) + nr_pages*sizeof (struct scatterlist);
+	/*
+	 * If the maximum block size increases, inline sgl might not fit into
+	 * a single page. We might want to consider using chained sgl if
+	 * that's the case.
+	 */
+	ASSERT(nr_pages * sizeof (struct scatterlist) <= PAGE_SIZE);
+#ifndef DEBUG_ABD
+	abd = kmem_alloc(asize, KM_PUSHPAGE);
+#else
+	abd = umem_alloc_aligned(asize, PAGE_SIZE, UMEM_DEFAULT);
+	/* deny access to padding */
+	if (mprotect(abd, PAGE_SIZE, PROT_NONE) != 0) {
+		perror("mprotect failed");
+		ASSERT(0);
+	}
+#endif
+	ASSERT(abd);
+
+	return (abd);
+}
+
+static inline void
+abd_free_struct(abd_t *abd, int nr_pages)
+{
+#ifndef DEBUG_ABD
+	kmem_free(abd, sizeof (abd_t) + nr_pages*sizeof (struct scatterlist));
+#else
+	if (mprotect(abd, PAGE_SIZE, PROT_READ|PROT_WRITE) != 0) {
+		perror("mprotect failed");
+		ASSERT(0);
+	}
+	umem_free(abd, sizeof (abd_t) + nr_pages*sizeof (struct scatterlist));
+#endif
+}
+
+/*
+ * Allocate a new ABD to point to offset @off of the original ABD.
+ * It shares the underlying buffer with the original ABD.
+ * Use abd_put to free. The original ABD(allocated from abd_alloc) must
+ * not be freed before any of its derived ABD.
+ */
+abd_t *
+abd_get_offset(abd_t *sabd, size_t off)
+{
+	size_t offset;
+	abd_t *abd;
+
+	ABD_CHECK(sabd);
+	ASSERT(off <= sabd->abd_size);
+
+	abd = abd_alloc_struct(0);
+
+	abd->abd_magic = ARC_BUF_DATA_MAGIC;
+	abd->abd_size = sabd->abd_size - off;
+
+	if (ABD_IS_LINEAR(sabd)) {
+		abd->abd_flags = ABD_F_LINEAR;
+		abd->abd_offset = 0;
+		abd->abd_nents = 1;
+		abd->abd_buf = sabd->abd_buf + off;
+	} else {
+		abd->abd_flags = ABD_F_SCATTER;
+		offset = sabd->abd_offset + off;
+		abd->abd_offset = offset & (PAGE_SIZE - 1);
+		/* make sure the new abd start as sgl[0] */
+		abd->abd_sgl = &sabd->abd_sgl[offset >> PAGE_SHIFT];
```

Comment:
@tuxoko 
Since the buildbots are upset about every itty-bitty thing and your patchset, due to the "pad" change are tested more on #3189 , #3190 I'd wanted to share the results with you:

http://buildbot.zfsonlinux.org/builders/ubuntu-12.10-amd64-builder/builds/2883/steps/shell_4
http://buildbot.zfsonlinux.org/builders/ubuntu-12.10-amd64-builder/builds/2883/steps/shell_4/logs/stdio

```
../../module/zfs/abd.c: In function 'abd_get_offset':
../../module/zfs/abd.c:975:43: error: 'PAGE_SHIFT' undeclared (first use in this function)
../../module/zfs/abd.c:975:43: note: each undeclared identifier is reported only once for each function it appears in
../../module/zfs/abd.c: In function 'abd_alloc_scatter':
../../module/zfs/abd.c:1027:38: error: 'PAGE_SHIFT' undeclared (first use in this function)
make[5]: *** [abd.lo] Error 1
```

not sure how to tackle this - since:

PAGE_SHIFT is defined at
include/asm-generic/page.h
and subsequently for each arch on an option-basis differently, but default seems to be 12
arch/*/include/asm/page.h